src/libcore/borrow.rs

```diff
@@ -12,26 +12,143 @@
+
+/// A trait identifying how borrowed data behaves.
+///
+/// If a type implements this trait, it signals that a reference to it behaves
+/// exactly like a reference to `Borrowed`. As a consequence, if a trait is
+/// implemented both by `Self` and `Borrowed`, all trait methods that
+/// take a `&self` argument must produce the same result in both
+/// implementations.
+///
+/// As a consequence, this trait should only be implemented for types managing
+/// a value of another type without modifying its behavior. Examples are
+/// smart pointers such as [`Box`] or [`Rc`] as well the owned version of
+/// slices such as [`Vec`].
+///
+/// A relaxed version that allows providing a reference to some other type
+/// without any further promises is available through [`AsRef`].
+///
+/// When writing generic code, a use of `Borrow` should always be justified
+/// by additional trait bounds, making it clear that the two types need to
+/// behave identically in a certain context. If the code should merely be
+/// able to operate on any type that can produce a reference to a given type,
+/// you should use [`AsRef`] instead.
+///
+/// The companion trait [`BorrowMut`] provides the same guarantees for
+/// mutable references.
+///
+/// [`Box`]: ../boxed/struct.Box.html
+/// [`Rc`]: ../rc/struct.Rc.html
+/// [`Vec`]: ../vec/struct.Vec.html
+/// [`AsRef`]: ../convert/trait.AsRef.html
+/// [`BorrowMut`]: trait.BorrowMut.html
+///
+/// # Examples
+///
+/// As a data collection, [`HashMap`] owns both keys and values. If the key’s
+/// actual data is wrapped in a managing type of some kind, it should,
+/// however, still be possible to search for a value using a reference to the
+/// key’s data. For instance, if the key is a string, then it is likely
+/// stored with the hash map as a [`String`], while it should be possible
+/// to search using a [`&str`][`str`]. Thus, `insert` needs to operate on a
+/// string while `get` needs to be able to use a `&str`.
+///
+/// Slightly simplified, the relevant parts of `HashMap` look like this:
+///
+/// ```
+/// use std::borrow::Borrow;
+/// use std::hash::Hash;
+///
+/// pub struct HashMap<K, V> {
+///     # marker: ::std::marker::PhantomData<(K, V)>,
+///     // fields omitted
+/// }
+///
+/// impl<K, V> HashMap<K, V> {
+///     pub fn insert(&self, key: K, value: V) -> Option<V>
+///         where K: Hash + Eq
+///     {
+///         # unimplemented!()
+///         // ...
+///     }
+///
+///     pub fn get<Q>(&self, k: &Q) -> Option<&V>
+///         where K: Borrow<Q>,
+///               Q: Hash + Eq + ?Sized
+///     {
+///         # unimplemented!()
+///         // ...
+///     }
+/// }
+/// ```
+///
+/// The entire hash map is generic over the stored type for the key, `K`.
+/// When inserting a value, the map is given such a `K` and needs to find
+/// the correct hash bucket and check if the key is already present based
+/// on that `K` value. It therefore requires `K: Hash + Eq`.
```

Comment:
"that `K` value" is a bit confusing since it mixes up "value" as in "data" with "value" as in "`V`".